zig: build ci

matklad opened the pull request:
Every time I subimt a PR, I fail either `zig fmt`, `zig build check`, or `zig build tidy`. I of course remember which one failed, and run that locally before the next PR. Sadly, I can remeber running two commands at most, and it is always the third one that fails!

So, let's introduce `zig build ci` which runs a reasonable set of checks.

matklad committed:
build: don't touch me

matklad committed:
zig: build ci

Every time I subimt a PR, I fail either `zig fmt`, `zig build check`, or
`zig build tidy`. I of course remember which one failed, and run that
locally before the next PR. Sadly, I can remeber running two commands at
most, and it is always the third one that fails!

So, let's introduce `zig build ci` which runs a reasonable set of
checks.

cb22 commented on zig/download.ps1:
```diff
@@ -0,0 +1,6 @@
+#!/bin/sh
+echo `# <#`
+./zig/download.sh
+exit
+#> > $null
+./zig/download.bat
```
I'm not quite sure if I love this or hate this :smile:.

matklad replied:
I mean, yes, but also consider that the alternative is adding ifs to YAML, which I'd argue is as elegant as putting bash inside powershell :P

cb22 commented on .github/workflows/ci.yml:
```diff
@@ -20,10 +20,27 @@ jobs:
+        include:
+          - { os: 'ubuntu-latest' }
+          - { os: 'ubuntu-latest-arm64' }
+          - { os: 'windows-latest' }
```
Out-of-scope for PR, but perhaps we should add Windows ARM64 too.

matklad replied:
We don't support building for arm windows, and I'd rather add that only when someone asks. I am confident that we _could_ support whatever, but I want to minimize the time to administrate CI heterogeneity (my mental model is that any CI config has an independent probability of "flakes" for boring reasons).

cb22 commented on .github/workflows/ci.yml:
```diff
@@ -20,10 +20,27 @@ jobs:
+    steps:
+      - run: git config --global core.autocrlf false
+      - if: matrix.os == 'ubuntu-latest' || matrix.os == 'ubuntu-latest-arm64'
+        run: | # Allow unsare for vortex.
```
```suggestion
        run: | # Allow unshare for vortex.
```

cb22 commented on .github/workflows/ci.yml:
```diff
@@ -225,7 +178,7 @@ jobs:
     if: always() && github.event_name == 'merge_group'
     runs-on: ubuntu-latest
-    needs: [smoke, test_alpine, test_ubuntu, test_aof, test_windows, test_macos, clients, devhub]
+    needs: [smoke, test, test_alpine, test_aof, clients, devhub]
```
Would be nice if we could somehow drop `test_alpine` :thinking:.

matklad replied:
I am confident that we can just drop it, which I did.

cb22 commented on build.zig:
```diff
@@ -359,6 +366,147 @@ fn build_vsr_module(b: *std.Build, options: struct {
+}
+
+fn build_ci_check_status(step: *std.Build.Step.Run) void {
+    // XXX: The purpose here is not to check status (Zig does that automatically), but rather to
```
A bit of _careful_ Googling tells me `XXX:` is indeed used as a heads-up (I've always used it more as a `TODO`). That said, we don't use it anywhere else, so perhaps `NB: `?

matklad replied:
Yeah, let's use NB as more classy. They have semantic distinction of 

* NB: something unusual happens here
* XXX: something unusual happens here which will bite you in the back if you don't pay attention

but that's perhaps not worth it.

matklad committed:
ci: drop test alpine

Database is statically linked, so alpine doesn't add any useful
dimensionality here. Alpine is useful for cilents which are dynamically
linked, and we have _some_ tests for that in node and dotnet ci.zig.
